storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -2826,9 +2915,22 @@ public void launchServer() throws Exception {
                 .parallelStream()
                 .mapToDouble(SupervisorResources::getTotalCpu)
                 .sum());
-
+            StormMetricsRegistry.registerGauge("nimbus:longest-scheduling-time-ms", () -> {
+                Long currTime = Time.nanoTime();
+                Long startTime = schedulingStartTime.get();
+                //There could be race condition here but seems trivial, elapsed is
+                // guaranteed to be no longer than real elapsed time of scheduling
+                Long longest = longestSchedulingTime.get();
+                if (startTime != null) {
+                    longest = currTime - startTime > longest ? currTime - startTime : longest;
+                }
+                //To millis. How should I put the constant for magic numbers?
```

Comment:
I'm concerned of loss of precision here since all of them return long instead of double after conversion.

Reply:
Why would there be a loss of precision? The input number is a long, which is an integral type (not floating point).

Reply:
Sorry, I get what you're saying. Disregard the above. If you're making the gauge measure milliseconds it probably still makes sense to use whole milliseconds rather than a floating point representation.

Reply:
Because the decimals are too trivial?

Reply:
Yes, or if the decimals are meaningful the gauge should be using nanoseconds instead of millis.